cm_ovirt_vm_creator.py

```diff
@@ -143,46 +143,83 @@ def construct_search_by_name_query(node_name):
     return search_string.format(node_name=node_name).__str__()
 
 
+def chunks(l, n):
+    """Yield successive n-sized chunks from l."""
+    for i in range(0, len(l), n):
+        yield l[i:i + n]
+
+
+def start_vm(vm_service, args):
+    """ Start a VM """
+    pub_sshkey = os.environ[args.pub_sshkey]
+    # ovirt API quirk - starting a VM with wait=False does nothing,
+    # and unlike other APIs, it accept async=True to make it asynchronous
+    vm_service.start(async=True,
+                     use_cloud_init=True,
+                     vm=types.Vm(initialization=types.Initialization(authorized_ssh_keys=pub_sshkey)))
+
+
 def create_vms(cluster_nodes, args):
     """ creates the vms in cluster_nodes list, and skipps if they exist """
     vm_dict = {}
-    for node in cluster_nodes:
-        print("node=%s" % (node), file=sys.stderr)
-        tmp = vms_service.list(search=construct_search_by_name_query(node))
-        if len(tmp) == 1:
-            vm_dict[node] = vms_service.vm_service(tmp[0].id)
-            print("VM %s was found ... skipping creation" % (node), file=sys.stderr)
-        else:
-            vm = vms_service.add(types.Vm(name=node,
-                                          cluster=types.Cluster(name=args.ovirt_cluster),
-                                          template=types.Template(name=args.ovirt_template)))
+    for block in chunks(cluster_nodes, args.block_size):
+        block_vm_futures = []
+        for node in block:
+            print("node=%s" % (node), file=sys.stderr)
+            tmp = vms_service.list(search=construct_search_by_name_query(node))
+            if len(tmp) == 1:
+                vm_dict[node] = vms_service.vm_service(tmp[0].id)
+                # start VM if not already started
+                vm = vm_dict[node].get()
+                if vm.status == types.VmStatus.DOWN:
+                    start_vm(vm_dict[node], args)
+                print("VM %s was found ... skipping creation" % (node), file=sys.stderr)
+            else:
+                vm_future = vms_service.add(types.Vm(name=node,
+                                                     cluster=types.Cluster(name=args.ovirt_cluster),
+                                                     template=types.Template(name=args.ovirt_template)), wait=False)
+                block_vm_futures.append((node, vm_future))
+
+        # wait for VMs of this block to be created, and then start each one
+        for node_name, future_vm in block_vm_futures:
```

Comment:
This loop should only call `future_vm.wait()`, otherwise some of the requests may not be sent, because you are doing other things. If needed, add another loop with the rest of the logic.

Reply:
I'm not sure I fully understand, why wouldn't the requests be sent?

Reply:
The asynchronous support in the oVirt Python SDK does not use multiple threads; everything happens in the thread that does calls to the SDK. Sending requests and receiving responses also happens in that thread, when you call `future.wait()`. If you are doing other things, then there will be no progress in the pending requests. In particular, if you call `time.sleep(...)` the thread will just do nothing, so no requests will be sent or responses received during that time.

Reply:
The reason I was calling `time.sleep` was not to wait for the response, it was to avoid overloading the server with too many requests in a short timeframe.

So if I understand correctly, what I need to do is to create all the VMs in a block, then wait on all VMs in the block, then sleep, then continue to the next block,

and then when all VMs are created start all of them in one loop, then wait on all the start futures in another loop (or maybe start them in blocks with sleeping in between each block, if we want to be extra careful the server won't be too overloaded).

Correct me if I'm wrong.

Reply:
Yes, I think that is a good approach. It may not be worth to do things in blocks, as you will not be creating or starting hundreds of virtual machines.